src/Http/Http.Extensions/src/RequestExtensions.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Http.Features;
+
+namespace Microsoft.AspNetCore.Http
+{
+    /// <summary>
+    /// Extensions for HttpRequest.
+    /// </summary>
+    public static class RequestExtensions
+    {
+        /// <summary>
+        /// Indicates if the request may have a body. Returns null if the server has not implemented this feature.
+        /// </summary>
+        /// <remarks>
+        /// This returns true when:
+        /// - It's an HTTP/1.x request with a non-zero Content-Length or a 'Transfer-Encoding: chunked' header.
+        /// - It's an HTTP/2 request that did not set the END_STREAM flag on the initial headers frame.
+        /// The final request body length may still be zero for the chunked or HTTP/2 scenarios.
+        /// This returns false when:
+        /// - It's an HTTP/1.x request with no Content-Length or 'Transfer-Encoding: chunked' header, or the Content-Length is 0.
+        /// - It's an HTTP/2 request that set END_STREAM on the initial headers frame.
+        /// When false, the request body should never return data.
+        /// This returns null if the underlying server has not implemented the feature.
+        /// </remarks>
+        public static bool? CanHaveBody(this HttpRequest request)
```

Comment:
Can we make this a property on the HttpRequest instead? I don't want to pollute the HttpRequset with core concepts like this.